gunicorn/systemd.py

```diff
@@ -43,3 +44,39 @@ def listen_fds(unset_environment=True):
         os.environ.pop('LISTEN_FDS', None)
 
     return fds
+
+
+def sd_notify(state, unset_environment=False, debug=False):
+    """Send a notification to systemd. state is a string; see
+    the man page of sd_notify (http://www.freedesktop.org/software/systemd/man/sd_notify.html)
+    for a description of the allowable values.
+
+    If the unset_environment parameter is True, sd_notify() will unset
+    the $NOTIFY_SOCKET environment variable before returning (regardless of
+    whether the function call itself succeeded or not). Further calls to
+    sd_notify() will then fail, but the variable is no longer inherited by
+    child processes.
+
+    Normally this method silently ignores exceptions (for example, if the
+    systemd notification socket is not available) to allow applications to
+    function on non-systemd based systems. However, setting debug=True will
+    cause this method to raise any exceptions generated to the caller, to
+    aid in debugging."""
+
+    addr = os.getenv('NOTIFY_SOCKET')
+    if addr is None:
+        # not run in a service, just a noop
+        return
+    if unset_environment:
+        os.unsetenv('NOTIFY_SOCKET')
+    try:
+        sock = socket.socket(socket.AF_UNIX, socket.SOCK_DGRAM | socket.SOCK_CLOEXEC)
+        if addr[0] == '@':
+            addr = '\0' + addr[1:]
+        sock.connect(addr)
+        sock.sendall(state.encode('utf-8'))
```

Comment:
This is just a question, I'm not sure which way is better, but instead of encoding a str object every time we call this function, can't we just pass a bytes object directly? We could even add a `isinstance(state, bytes)` at the top of the function if we want to be safe.

Reply:
I'd leave it at a string interface, seems more logical to me. also, systemd already specifies it expects utf-8.

 allowing for bytes with a `isinstance` maybe makes sense, but then I'd need to worry about non-utf-8 bytes